### PR TITLE
feature: isCommonlyUsed password check not hardcoded #4018

### PR DESCRIPTION
**Pull request by WillardHu**
Signed-off-by: WillardHu <wei.hu@daocloud.io>

## What's the purpose of this PR

`isCommonlyUsed` password check not hardcoded

## Which issue(s) this PR fixes:
Fixes 4018

## Brief changelog

XXXXX

Follow this checklist to help us incorporate your contribution quickly and easily:

- [x] Read the [Contributing Guide](https://github.com/ctripcorp/apollo/blob/master/CONTRIBUTING.md) before making this pull request.
- [x] Write a pull request description that is detailed enough to understand what the pull request does, how, and why.
- [x] Write necessary unit tests to verify the code.
- [x] Run `mvn clean test` to make sure this pull request doesn't break anything.
- [ ] Update the [`CHANGES` log](https://github.com/ctripcorp/apollo/blob/master/CHANGES.md).

### CODE REVIEW

File: apollo-portal/src/main/java/com/ctrip/framework/apollo/portal/component/config/PortalConfig.java [hunk 1]
```diff
@@ -273,4 +275,12 @@ public String getAdminServiceAccessTokens() {
   public boolean supportSearchByItem() {
     return getBooleanProperty("searchByItem.switch", true);
   }
+  
+  public List<String> listOFCodeFragment() {
+    String[] value = getArrayProperty("auth.userPasswordChecker.listOFCodeFragment", null);
```
Comment (nobodyiam) [inline apollo-portal/src/main/java/com/ctrip/framework/apollo/portal/component/config/PortalConfig.java:280]: It may be better to change OF to Of?
BTW, there wasn't a good standard naming pattern for configurations. However, recently we thought it may look better if we separate the word with `-`, what do you think?
```suggestion
  public List<String> listOfCodeFragment() {
    String[] value = getArrayProperty("auth.user-password-checker.list-of-code-fragment", null);
```

File: apollo-portal/src/main/java/com/ctrip/framework/apollo/portal/util/checker/AuthUserPasswordChecker.java [hunk 2]
```diff
@@ -58,7 +65,13 @@ private boolean isCommonlyUsed(String password) {
     if (Strings.isNullOrEmpty(password)) {
       return true;
     }
-    for (String s : LIST_OF_CODE_FRAGMENT) {
+    List<String> listOfCodeFragment;
+    if (portalConfig != null && !portalConfig.listOFCodeFragment().isEmpty()) {
+      listOfCodeFragment = portalConfig.listOFCodeFragment();
+    } else {
+      listOfCodeFragment = DEFAULT_LIST_OF_CODE_FRAGMENT;
+    }
```
Comment (nobodyiam) [inline apollo-portal/src/main/java/com/ctrip/framework/apollo/portal/util/checker/AuthUserPasswordChecker.java:73]: portalConfig won't be null, so how about we do like this?
```suggestion
    List<String> listOfCodeFragment = portalConfig.listOFCodeFragment();
    if (listOfCodeFragment.isEmpty()) {
      listOfCodeFragment = DEFAULT_LIST_OF_CODE_FRAGMENT;
    }
```

File: apollo-portal/src/test/java/com/ctrip/framework/apollo/portal/util/AuthUserPasswordCheckerTest.java [hunk 3]
```diff
@@ -30,7 +30,7 @@
 
   @Before
   public void setup() {
-    checker = new AuthUserPasswordChecker();
+    checker = new AuthUserPasswordChecker(null);
```
Comment (nobodyiam) [inline apollo-portal/src/test/java/com/ctrip/framework/apollo/portal/util/AuthUserPasswordCheckerTest.java:33]: I think we need to pass a mock portal config here and test when it has `listOfCodeFragment`

File: apollo-portal/src/main/java/com/ctrip/framework/apollo/portal/component/config/PortalConfig.java (hunk 1 above)
Comment (WillardHu) [reply apollo-portal/src/main/java/com/ctrip/framework/apollo/portal/component/config/PortalConfig.java]: Done.

File: apollo-portal/src/main/java/com/ctrip/framework/apollo/portal/util/checker/AuthUserPasswordChecker.java (hunk 2 above)
Comment (WillardHu) [reply apollo-portal/src/main/java/com/ctrip/framework/apollo/portal/util/checker/AuthUserPasswordChecker.java]: Done.

File: apollo-portal/src/test/java/com/ctrip/framework/apollo/portal/util/AuthUserPasswordCheckerTest.java [hunk 4]
```diff
@@ -63,6 +58,8 @@ public void testRegexMatch() {
 
   @Test
   public void testIsWeakPassword() {
+    AuthUserPasswordChecker checker = new AuthUserPasswordChecker(PortalConfigFaker.empty());
```
Comment (nobodyiam) [inline apollo-portal/src/test/java/com/ctrip/framework/apollo/portal/util/AuthUserPasswordCheckerTest.java:61]: Is it better to use mockito to handle the mock part? e.g.
```suggestion
    PortalConfig portalConfig = Mockito.mock(PortalConfig.class);
    Mockito.when(portalConfig.listOfCodeFragment()).thenReturn(Collections.emptyList());
    AuthUserPasswordChecker checker = new AuthUserPasswordChecker(portalConfig);
```

File: apollo-portal/src/main/java/com/ctrip/framework/apollo/portal/component/config/PortalConfig.java [hunk 5]
```diff
@@ -273,4 +275,12 @@ public String getAdminServiceAccessTokens() {
   public boolean supportSearchByItem() {
     return getBooleanProperty("searchByItem.switch", true);
   }
+  
+  public List<String> listOfCodeFragment() {
+    String[] value = getArrayProperty("auth.user-password-checker.list-of-code-fragment", null);
```
Comment (Anilople) [inline apollo-portal/src/main/java/com/ctrip/framework/apollo/portal/component/config/PortalConfig.java:280]: Add the prefix to the key?
```suggestion
    String[] value = getArrayProperty("apollo.portal.auth.user-password-checker.list-of-code-fragment", null);
```
Alternative
* auth.user-password-checker.list-of-code-fragment
* apollo.portal.auth.user-password-checker.list-of-code-fragment
* apollo.portal.auth.user-password-fragment
* apollo.portal.auth.user-password-checker-fragment
* apollo.portal.auth.user-password-checker-code-fragment
* apollo.portal.auth.user-password-not-allow-list
* ...

An better key let user more understandable this feature.

File: apollo-portal/src/test/java/com/ctrip/framework/apollo/portal/util/AuthUserPasswordCheckerTest.java (hunk 4 above)
Comment (WillardHu) [reply apollo-portal/src/test/java/com/ctrip/framework/apollo/portal/util/AuthUserPasswordCheckerTest.java]: Done.

File: apollo-portal/src/main/java/com/ctrip/framework/apollo/portal/component/config/PortalConfig.java (hunk 5 above)
Comment (WillardHu) [reply apollo-portal/src/main/java/com/ctrip/framework/apollo/portal/component/config/PortalConfig.java]: Done.

File: apollo-portal/src/main/java/com/ctrip/framework/apollo/portal/util/checker/AuthUserPasswordChecker.java [hunk 6]
```diff
@@ -28,7 +29,7 @@
   private static final Pattern PWD_PATTERN = Pattern
       .compile("^(?=.*[0-9].*)(?=.*[a-zA-Z].*).{8,20}$");
 
-  private static final List<String> LIST_OF_CODE_FRAGMENT = Arrays.asList(
+  private static final List<String> DEFAULT_USER_PASSWORD_NOT_ALLOW_LIST = Arrays.asList(
```
Comment (Anilople) [inline apollo-portal/src/main/java/com/ctrip/framework/apollo/portal/util/checker/AuthUserPasswordChecker.java:32]: Move the default value to `PortalConfig`'s method?

```java
  private static final List<String> DEFAULT_USER_PASSWORD_NOT_ALLOW_LIST = Arrays.asList(xxx);

  public List<String> getUserPasswordNotAllowList() {
    String[] value = getArrayProperty("apollo.portal.auth.user-password-not-allow-list", null);
    if (value == null || value.length == 0) {
      return DEFAULT_USER_PASSWORD_NOT_ALLOW_LIST;
    }
    return Arrays.asList(value);
  }
```

File: apollo-portal/src/test/java/com/ctrip/framework/apollo/portal/util/AuthUserPasswordCheckerTest.java [hunk 7]
```diff
@@ -81,4 +80,24 @@ public void testIsWeakPassword() {
     Assert.assertTrue(res.isSuccess());
   }
 
+  @Test
+  public void testIsWeakPassword2() {
```
Comment (Anilople) [inline apollo-portal/src/test/java/com/ctrip/framework/apollo/portal/util/AuthUserPasswordCheckerTest.java:84]: Maybe we can add more test case like return empty list, single element list?

File: apollo-portal/src/main/java/com/ctrip/framework/apollo/portal/util/checker/AuthUserPasswordChecker.java (hunk 6 above)
Comment (WillardHu) [reply apollo-portal/src/main/java/com/ctrip/framework/apollo/portal/util/checker/AuthUserPasswordChecker.java]: Done.

File: apollo-portal/src/test/java/com/ctrip/framework/apollo/portal/util/AuthUserPasswordCheckerTest.java (hunk 7 above)
Comment (WillardHu) [reply apollo-portal/src/test/java/com/ctrip/framework/apollo/portal/util/AuthUserPasswordCheckerTest.java:107]: Done.

File: apollo-portal/src/main/java/com/ctrip/framework/apollo/portal/component/config/PortalConfig.java [hunk 8]
```diff
@@ -47,10 +47,21 @@
...
+      "0987", "9876", "8765", "7654", "6543", "5432", "4321", "3210",
+      "1q2w", "2w3e", "3e4r", "5t6y", "abcd", "qwer", "asdf", "zxcv"
+  );
+  
   /**
    * meta servers config in "PortalDB.ServerConfig"
    */
   private static final Type META_SERVERS = new TypeToken<Map<String, String>>(){}.getType();
+  
+
```
Comment (nobodyiam) [inline apollo-portal/src/main/java/com/ctrip/framework/apollo/portal/component/config/PortalConfig.java:64]: These 2 empty lines looks unnecessary.
```suggestion
  
  
```

Comment (WillardHu) [reply apollo-portal/src/main/java/com/ctrip/framework/apollo/portal/component/config/PortalConfig.java]: Done.

**Review by nobodyiam — APPROVED** [review]
Looks good to me now. @Anilople would you please help to take a look from your perspective?

**Review by Anilople — APPROVED** [review]
